cosmwasm/enclaves/shared/contract-engine/src/contract_validation.rs

```diff
@@ -188,6 +201,75 @@ pub fn validate_contract_key(
     }
 }
 
+/// validate_contract_key validates the contract key against the contract address and code hash. If the contract was previously migrated, it also validates the contract key proof against the original contract key.
+pub fn validate_contract_key(
+    base_env: &BaseEnv,
+    canonical_contract_address: &CanonicalAddr,
+    contract_code: &ContractCode,
+) -> Result<(), EnclaveError> {
+    let og_contract_key: [u8; CONTRACT_KEY_LENGTH] = base_env.get_og_contract_key()?;
+
+    if base_env.was_migrated() {
+        trace!("Contract was migrated, validating proof");
+
+        let current_contract_key: [u8; CONTRACT_KEY_LENGTH] =
+            base_env.get_current_contract_key()?;
+
+        validate_current_contract_key(
+            &current_contract_key,
+            canonical_contract_address,
+            contract_code,
+            Some(&og_contract_key),
+        )?;
+
+        let sent_contract_key_proof = base_env.get_current_contract_key_proof()?;
```

Comment:
I think I'm missing something here. What does the key proof add to the security of the execution? Seems to me that it's checking exactly the same parameters as the above `validate_current_contract_key`.

Reply:
@assafmo